### PR TITLE
add-detector-code

### PR DESCRIPTION
**Pull request by calvix**
add base bad node detector code
towards https://github.com/giantswarm/giantswarm/issues/10357, build from https://github.com/giantswarm/aws-operator/pull/2546

this is the main code for a library that is used for detecting bad nodes that will be terminated

it should run in each provider-specific operator and it will return a list of nodes which should be terminated

This  function should be executed in a reconciliation loop for each cluster (some cluster CR)

There are a few configuration options:
```
    // MaxNodeTerminationPercentage defines a maximum percentage of nodes that will be returned as 'marked for termination'
	// ie: if the value is 0.5 and cluster have 10 nodes, than `DetectBadNodes`can only return maximum of 5 nodes
	// marked for termination at single run
	MaxNodeTerminationPercentage float64
	// NotReadyTickThreshold defines a how many times the node must bee seen as NotReady in order to return it as 'marked for termination'
	NotReadyTickThreshold int
	// PauseBetweenTermination defines a pause between 2 intervals where node termination can occur.
	// This is a safeguard to prevent nodes being terminated over and over or to not terminate too much at once.
	// ie: if the value is 5m it means once it returned nodes for termination it wont return another nodes for another 5 min.
	PauseBetweenTermination time.Duration
```

All have some sane defaults if not defined.

implementation of `PauseBetweenTermination` will be seaprate PR to have the scope of the PR small

### CODE REVIEW

**Review by xh3b4sd — COMMENTED** [review]
I think this kind of logic deserves unit tests. Before writing them we could maybe check to decouple logic and transmission, so that the `k8sClient` calls to the `apiserver` are not buried somewhere. Then we would not have to deal with some fancy mocking in the unit tests.

File: pkg/detector/detector.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+		}
+	}
+
+	nodesToTerminate, err := d.detectBadNodes(ctx, nodeList.Items)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// remove additional master nodes to avoid multiple master node termination at the same time
+	nodesToTerminate = d.removeMultipleMasterNodes(ctx, nodesToTerminate)
```
Comment (whites11) [inline pkg/detector/detector.go:97]: what if the failed node is the only master for a cluster?
Do we want to terminate it anyway?

File: pkg/detector/detector.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+func (d *Detector) maximumNodeTermination(nodeCount int) int {
+	limit := math.Round(float64(nodeCount) * d.maxNodeTerminationPercentage)
+
+	if limit < 1 {
+		limit = 1
+	}
+	return int(limit)
+}
+
+// removeMultipleMasterNodes removes multiple master nodes from the list to avoid more than 1 master node termination at same time
```
[the remark this answers is not in the document]
Comment (whites11) [reply pkg/detector/detector.go:198]: I think from the second master node in `nodeList` on, the code flow will go in the `else` branch at line 208 and thus all but the first master node are not added in the `filteredNodes` slice

**Review by whites11 — APPROVED** [review]
It seems reasonable to me.

File: pkg/detector/detector.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	}
+	if config.PauseBetweenTermination == 0 {
+		config.PauseBetweenTermination = defaultPauseBetweenTermination
+	}
+
+	d := &Detector{
+		logger:    config.Logger,
+		k8sClient: config.K8sClient,
+
+		notReadyTickThreshold:   config.NotReadyTickThreshold,
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:74]: fixed

File: pkg/detector/detector.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	}
+	return badNodes, nil
+}
+
+// nodeNotReady returns true of the node is not ready for certain period of time
+// this is used to detect bad nodes
+func nodeNotReady(n corev1.Node) bool {
+	for _, c := range n.Status.Conditions {
+		// find kubelet "ready" condition
+		if c.Type == "Ready" && c.Status != "True" {
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:133]: done

File: pkg/detector/detector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:200]: the returned list should contain maximum of 1 master node and unlimited number of worker nodes

the `foundMasterNode` bool there should make sure that only 1 master is appended to the `filteredNodes` array

this definetly deserves a test unit

File: pkg/detector/detector.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+	return nodesToTerminate, nil
+}
+
+// detectBadNodes looks at all nodes and adds a tick counter on each node if it sees it as NotReady
+// It will return all nodes which reached threshold for the tick counter.
+func (d *Detector) detectBadNodes(ctx context.Context, nodes []corev1.Node) ([]corev1.Node, error) {
+	var badNodes []corev1.Node
+	for _, n := range nodes {
+		//
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:117]: cleared

File: pkg/detector/detector.go (hunk 1 above)
Comment (calvix) [reply pkg/detector/detector.go:99]: If the master node is not ready then it kinda makes sense as some important components like `kiam-server` are running on master node and should be running

But we can have a bigger discussion on this

File: pkg/detector/detector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:200]: done, a unit test is there and it works as expected

**Comment by calvix** [comment]
RFR @xh3b4sd

File: pkg/detector/detector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply pkg/detector/detector.go:194]: Can the we improve the comments? There is not a single word about worker nodes and how the returned list looks like or tries to accomplish.

**Review by xh3b4sd — APPROVED** [review]
The comments are a bit bumpy. I like the unit tests. When this is how it should be from your perspective then let's go.

File: pkg/detector/detector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/detector/detector.go:194]: added few more comments

**Pull request merged**